onnxruntime/core/providers/xnnpack/nn/max_pool.cc

```diff
@@ -0,0 +1,193 @@
+  uint32_t stride_height = gsl::narrow<uint32_t>(pool_attrs_.strides[0]);
+  uint32_t stride_width = gsl::narrow<uint32_t>(pool_attrs_.strides[1]);
+  uint32_t dilation_height = gsl::narrow<uint32_t>(pool_attrs_.dilations[0]);
+  uint32_t dilation_width = gsl::narrow<uint32_t>(pool_attrs_.dilations[0]);
+
+  // get values from any fusion with an activation
+  std::string activation;
+  if (info.GetAttr<std::string>("activation", &activation).IsOK()) {
+    if (activation == "Clip" || activation == "Relu") {
+      std::vector<float> activation_params;
+
+      // min/max could be from Clip or Relu
+      if (info.GetAttrs<float>("activation_params", activation_params).IsOK()) {
+        if (activation_params.size() == 2) {
+          clip_min_max_ = {activation_params[0], activation_params[1]};
+        }
+      }
+    }
+  }
+
+  float output_min = clip_min_max_ ? clip_min_max_->first : -INFINITY;
+  float output_max = clip_min_max_ ? clip_min_max_->second : INFINITY;
+
+  uint32_t flags = 0;
+  if (pool_attrs_.auto_pad == AutoPadType::SAME_UPPER) {
+    flags |= XNN_FLAG_TENSORFLOW_SAME_PADDING;
+  }
+
+  // input is NHWC and we only support input with 4 dims. we checked C, H, W were all known in the op support checker
+  const auto& X_arg = *Node().InputDefs()[0];
+  const auto& X_shape = *X_arg.Shape();
+  int64_t H = X_shape.dim(1).dim_value();
+  int64_t W = X_shape.dim(2).dim_value();
+  int64_t C = X_shape.dim(3).dim_value();
+
+  // create NCHW shape to calculate most of the output shape. 'N' is set in Compute.
+  TensorShapeVector input_shape{1, C, H, W};
+  auto pads = pool_attrs_.pads;
+  auto nchw_output_dims = pool_attrs_.SetOutputSize(input_shape, C, &pads);
+  output_dims_ = {-1, nchw_output_dims[2], nchw_output_dims[3], nchw_output_dims[1]};
+
+  // TEMPORARY sanity check. If C, H and W are known, the output shape should have been able to be inferred, with the
+  // exception of the batch size. Can be removed once we've run more models using xnnpack MaxPool.
+  auto inferred_output_shape = utils::GetTensorShapeFromTensorShapeProto(*Node().OutputDefs()[0]->Shape());
+  ORT_ENFORCE(inferred_output_shape[1] == output_dims_[1] &&
+                  inferred_output_shape[2] == output_dims_[2] &&
+                  inferred_output_shape[3] == output_dims_[3],
+              "Shape mismatch between inferred value and calculated value.");
+
+  xnn_status status;
+  struct xnn_operator* p;
+  status = xnn_create_max_pooling2d_nhwc_f32(input_padding_top, input_padding_right,
+                                             input_padding_bottom, input_padding_left,
+                                             pooling_height, pooling_width,
+                                             stride_height, stride_width,
+                                             dilation_height, dilation_width,
+                                             C, C, C,  // channels, input_pixel_stride, output_pixel_stride
```

Comment:
why are they all C?

Reply:
I believe that's the normal usage.

e.g. 
https://github.com/google/XNNPACK/blob/802a217336706bc8e2ac6ce254569c0e34074926/bench/max-pooling.cc#L133